FIAT/guzman_neilan.py

```diff
@@ -30,29 +43,68 @@ def ExtendedGuzmanNeilanSpace(ref_el, subdegree, reduced=False):
         B = modified_bubble_subspace(B)
 
     if reduced:
-        BR = BernardiRaugel(ref_el, sd, subdegree=subdegree).get_nodal_basis()
+        BR = BernardiRaugel(ref_el, order).get_nodal_basis()
         reduced_dim = BR.get_num_members() - (sd-1) * (sd+1)
         BR = BR.take(list(range(reduced_dim)))
     else:
-        BR = ExtendedBernardiRaugelSpace(ref_el, subdegree)
+        BR = ExtendedBernardiRaugelSpace(ref_el, order)
+
     GN = constant_div_projection(BR, C0, B)
+    if kind == 2:
+        Bk = take_interior_bubbles(C0, order)
+        GN = polynomial_set.polynomial_set_union_normalized(GN, Bk)
     return GN
 
 
 class GuzmanNeilan(finite_element.CiarletElement):
-    """The Guzman-Neilan extended element."""
-    def __init__(self, ref_el, degree=None, subdegree=1):
+    """The Guzman-Neilan (extended) element of the first kind.
+
+    Reference element: a simplex of any dimension.
+    Function space: Pk^d + normal facet bubbles with div in P0, with 1 <= k < dim.
+    Degrees of freedom: evaluation at Pk lattice points, and normal moments on faces.
+
+    This element belongs to a Stokes complex, and is paired with unsplit DG_{k-1}.
+    """
+    def __init__(self, ref_el, order=1):
         sd = ref_el.get_spatial_dimension()
-        if degree is None:
-            degree = sd
-        if degree != sd:
-            raise ValueError("Guzman-Neilan only defined for degree = dim")
-        poly_set = ExtendedGuzmanNeilanSpace(ref_el, subdegree)
-        dual = BernardiRaugelDualSet(ref_el, degree, subdegree=subdegree)
+        if order >= sd:
+            raise ValueError(f"{type(self).__name__} is only defined for order < dim")
+        degree = sd
+        poly_set = ExtendedGuzmanNeilanSpace(ref_el, order)
+        dual = BernardiRaugelDualSet(ref_el, order, degree=degree)
         formdegree = sd - 1  # (n-1)-form
         super().__init__(poly_set, dual, degree, formdegree, mapping="contravariant piola")
 
 
+class GuzmanNeilanSecondKind(finite_element.CiarletElement):
```

Comment:
If you have a second kind element, shouldn't you explicitly call the other one first kind in the name?

Reply:
In FIAT/FInAT we have Nedelec and NeledecSecondKind classes. The element names we register in UFL are "Nedelec First Kind H(curl)" and "Nedelec Second Kind H(curl)". 

We don't really need two separate classes for GuzmanNeilan in FIAT, just a `kind` kwarg, FInAT may have two separate clases, just as I did for HCT and HCT-red.

Then I propose to register the names
```
"Guzman-Neilan First Kind H1" 
"Guzman-Neilan Second Kind H1"
"Guzman-Neilan H1(div)" 
```

Reply:
Alternatively, instead of distinguishing them with kind, we could reuse the mechanism that would make Lagrange get the extra Lagrange basis functions that are added by Guzman-Neilan Second Kind. Basically, 

```
GuzmanNeilanFirstKind(cell, order) = VectorElement(Lagrange(cell, order)) + GuzmanNeilanBubble(cell)
GuzmanNeilanSecondKind(cell, order) = VectorElement(Lagrange(cell, order, variant="alfeld")) + GuzmanNeilanBubble(cell)
```
 This suggests that the only thing FIAT should care about is GuzmanNeilanBubble. 
 
 That should be fine as long as we support enrichment of non-macro elements with macroelements

Reply:
We will have to wait to get the UFL element algebra, and use more descriptive element names